netket/hilbert/abstract_hilbert.py

```diff
@@ -0,0 +1,131 @@
+import abc
+import numpy as _np
+
+
+"""int: Maximum number of states that can be indexed"""
+max_states = (_np.iinfo(_np.intp).max)
+
+
+class AbstractHilbert(abc.ABC):
+    """Abstract class for NetKet hilbert objects"""
+
+    @property
+    @abc.abstractmethod
+    def size(self):
+        r"""int: The total number number of spins."""
+        return NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def is_discrete(self):
+        r"""bool: Whether the hilbert space is discrete."""
+        return NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def is_finite(self):
+        r"""bool: Whether the local hilbert space is finite."""
+        return NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def local_size(self):
+        r"""int: Size of the local degrees of freedom that make the total hilbert space."""
+        return NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def local_states(self):
+        r"""list[float]: A list of discreet local quantum numbers."""
+        return NotImplementedError
+
+    def numbers_to_states(self, numbers, out=None):
+        r"""Returns the quantum numbers corresponding to the n-th basis state
+        for input n. n is an array of integer indices such that numbers[k]=Index(states[k]).
+        Throws an exception iff the space is not indexable.
+        Args:
+            numbers: Batch of input numbers to be converted into arrays of quantum numbers.
+            out: Array of quantum numbers corresponding to numbers.
+                 If None, memory is allocated.
+        """
+        return NotImplementedError
+
+    def number_to_state(self, number):
+        r"""Returns the quantum number corresponding to the n-th basis state
+        for input n. n is a integer index such that number=Index(state).
+        Throws an exception iff the space is not indexable.
+        Args:
+            numbers: Batch of input numbers to be converted into arrays of quantum numbers.
+
+        Returns:
+            numpy.array: An array of quantum numbers corresponding to the state.
+        """
+        return self.numbers_to_states(_np.atleast_1d(number))
```

Comment:
This returns a 2D array. 
I think this should return a single state, so a 1D array?

```suggestion
        return self.numbers_to_states(_np.atleast_1d(number))[0,:]
```